runtime/include/pool.h

```diff
@@ -0,0 +1,102 @@
+                                                                                                                  \
+		obj = ps_list_head_first_d(&self->list, struct STRUCT_NAME);                                      \
+		assert(obj);                                                                                      \
+		ps_list_rem_d(obj);                                                                               \
+		LOCK_UNLOCK(&self->lock);                                                                         \
+		return obj;                                                                                       \
+	}                                                                                                         \
+                                                                                                                  \
+	static inline int STRUCT_NAME##_pool_add_nolock(struct STRUCT_NAME##_pool *self, struct STRUCT_NAME *obj) \
+	{                                                                                                         \
+		assert(self != NULL);                                                                             \
+		assert(obj != NULL);                                                                              \
+		assert(!self->use_lock || LOCK_IS_LOCKED(&self->lock));                                           \
+                                                                                                                  \
+		ps_list_head_add_d(&self->list, obj);                                                             \
+		return 0;                                                                                         \
+	}                                                                                                         \
+                                                                                                                  \
+	static inline int STRUCT_NAME##_pool_add(struct STRUCT_NAME##_pool *self, struct STRUCT_NAME *obj)        \
+	{                                                                                                         \
+		assert(self != NULL);                                                                             \
+		assert(obj != NULL);                                                                              \
+		assert(self->use_lock);                                                                           \
+                                                                                                                  \
+		LOCK_LOCK(&self->lock);                                                                           \
+		ps_list_head_add_d(&self->list, obj);                                                             \
+		LOCK_UNLOCK(&self->lock);                                                                         \
+		return 0;                                                                                         \
+	}
```

Comment:
I might be completely missing something here, but I don't see any type specialization that is necessary, nor any fast-paths that require aggressive inlining of the macro arguments. The DTOR could be a function pointer within the structure. Because of this, I don't see why this needs to be a macro rather than a simple generic data-structure. I'm not confident I'm understanding everything here, so I could be wrong.

I don't think it is relevant here, but if you need a lock-free cache for this stuff, https://github.com/gwsystems/ps/blob/master/ps_pgalloc.h